### PR TITLE
[ci-visibility] Support CI Visibility's manual API

### PR DESCRIPTION
**Pull request by juan-fernandez**
### What does this PR do?
* Create a `TestApiManualPlugin` to listen to newly created channels. 
* Add `DD_CIVISIBILITY_MANUAL_API_ENABLED` to enable manual api. 
* Add e2e tests. 

### Motivation
Allow customers who use a non supported test framework to use Datadog.

### CODE REVIEW

File: integration-tests/manual-api/test.fake.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+      childOf: testSpan
+    })
+    return new Promise((resolve) => {
+      setTimeout(() => {
+        childSpan.finish()
+        resolve()
+      }, 10)
+    })
+  })
+})
```
Comment (juan-fernandez) [inline integration-tests/manual-api/test.fake.js:48]: ℹ️ these are tests run by the "fake testing framework"

File: integration-tests/manual-api/test.fake.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
+/* eslint-disable */
```
Comment (juan-fernandez) [inline integration-tests/manual-api/test.fake.js:1]: ℹ️ these are tests run by the "fake testing framework"

File: packages/datadog-plugin-mocha/src/index.js [hunk 3]
```diff
@@ -70,7 +70,7 @@ class MochaPlugin extends CiPlugin {
     this.addSub('ci:mocha:test-suite:finish', (status) => {
```
Comment (juan-fernandez) [inline packages/datadog-plugin-mocha/src/index.js:70]: unrelated fixes below

File: ci/init.js [hunk 4]
```diff
@@ -44,6 +46,10 @@ if (isJestWorker) {
 if (shouldInit) {
   tracer.init(options)
   tracer.use('fs', false)
+  if (isTrue(process.env.DD_CIVISIBILITY_MANUAL_API_ENABLED)) {
+    // To fake that we're loading a "manual" library, which triggers the instantiation of ManualPlugin
```
Comment (rochdev) [inline ci/init.js:50]: Why is this needed?

File: ci/init.js [hunk 5]
```diff
@@ -44,6 +46,10 @@ if (isJestWorker) {
 if (shouldInit) {
   tracer.init(options)
   tracer.use('fs', false)
+  if (isTrue(process.env.DD_CIVISIBILITY_MANUAL_API_ENABLED)) {
```
Comment (rochdev) [inline ci/init.js:49]: Is this standardized across languages?

File: packages/dd-trace/src/plugins/index.js [hunk 6]
```diff
@@ -35,6 +35,7 @@ module.exports = {
   get 'https' () { return require('../../../datadog-plugin-http/src') },
   get 'ioredis' () { return require('../../../datadog-plugin-ioredis/src') },
   get 'jest-circus' () { return require('../../../datadog-plugin-jest/src') },
+  get 'manual' () { return require('../../../datadog-plugin-manual/src') },
```
Comment (rochdev) [inline packages/dd-trace/src/plugins/index.js:38]: Is this really a plugin though? It's not instrumenting a module, so it seems like a hack to me. This is especially true since it also needs a fake instrumentation event to be instantiated.

**Review by rochdev — CHANGES_REQUESTED** [review]

File: ci/init.js (hunk 5 above)
Comment (juan-fernandez) [reply ci/init.js:49]: it isn't: nodejs' is the only tracer doing this. I still think it's a good idea to do a first release behind a feature flag

File: packages/dd-trace/src/plugins/index.js (hunk 6 above)
Comment (juan-fernandez) [reply packages/dd-trace/src/plugins/index.js:38]: this is a different option for initialization: https://github.com/DataDog/dd-trace-js/pull/3202/commits/8bcbac88324a7bb8d81ef883be91f7478561f290

File: ci/init.js (hunk 5 above)
Comment (rochdev) [reply ci/init.js]: Oh yeah I meant the name of the environment variable. If it only applies to Node then it's fine 👍

File: packages/dd-trace/src/plugins/index.js (hunk 6 above)
Comment (rochdev) [reply packages/dd-trace/src/plugins/index.js:38]: > well I'd argue that it is instrumenting a module
Is it though? If I'm not misunderstanding, it seems like they would just emit the right events. How would it work to use this outside of CI too, especially with the generic `manual` name? The more I think about it the less I think this should be a plugin, or at least not the same kind of plugin than for instrumented modules.

File: integration-tests/manual-api/test.fake.js [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+  })
+  test('integration test', () => {
+    // Just for testing purposes, so we don't create a custom span
+    if (!process.env.DD_CIVISIBILITY_MANUAL_API_ENABLED) {
+      return Promise.resolve()
+    }
+    const testSpan = tracer.scope().active()
+    const childSpan = tracer.startSpan('custom.span', {
+      childOf: testSpan
+    })
```
Comment (juan-fernandez) [inline integration-tests/manual-api/test.fake.js:40]: @rochdev  I added `dd-trace:ci:manual:test:addTags` but left this bit with our current API. While this could be done via events, I don't want to prematurely expand the new manual api to be able to create arbitrary spans.

**Comment by juan-fernandez** [comment]
this should be ready for a review @rochdev

File: integration-tests/manual-api/test.fake.js (hunk 2 above)
Comment (rochdev) [reply integration-tests/manual-api/test.fake.js:1]: Can the folder be moved maybe to `ci`? `manual-api` really seems generic.

File: integration-tests/manual-api/test.fake.js (hunk 7 above)
Comment (rochdev) [reply integration-tests/manual-api/test.fake.js:40]: I'm not sure I understand the reasoning. Creating spans is a completely separate event, so adding an event to add tags here would not give the ability to create arbitrary spans. If you're worried about events, which would definitely be justified since it's the first time we introduce any publicly, then I would say to simply avoid events _entirely_ at that point, at least for now.

Comment (juan-fernandez) [reply integration-tests/manual-api/test.fake.js:40]: we might not be understanding each other. I meant I'm ok with `dd-trace:ci:manual:test:addTags`, and I added it, but I **didn't** add an extra `dd-trace:ci:manual:test:addCustomSpan`, which would have to receive, at least, `childOf`, `name` and `tags`. I think it's too flexible (and dangerous) for a first version of this API

File: integration-tests/manual-api/test.fake.js (hunk 2 above)
Comment (juan-fernandez) [reply integration-tests/manual-api/test.fake.js]: I moved it to `integration-tests/ci-visibility/test-api-manual`

**Comment by juan-fernandez** [comment]
is there any other comment @rochdev ?

File: integration-tests/ci-visibility/test-api-manual/run-fake-test-framework.js [hunk 8]
```diff
@@ -0,0 +1,26 @@
+/* eslint-disable */
+require('./setup-fake-test-framework')
```
Comment (uurien) [inline integration-tests/ci-visibility/test-api-manual/run-fake-test-framework.js:2]:
```suggestion
'use strict'

/* eslint-disable */

require('./setup-fake-test-framework')
```

File: integration-tests/ci-visibility/test-api-manual/setup-fake-test-framework.js [hunk 9]
```diff
@@ -0,0 +1,31 @@
+global.tests = []
```
Comment (uurien) [inline integration-tests/ci-visibility/test-api-manual/setup-fake-test-framework.js:1]:
```suggestion
'use strict'

global.tests = []
```

File: integration-tests/ci-visibility/test-api-manual/run-fake-test-framework.js (hunk 8 above)
Comment (uurien) [inline integration-tests/ci-visibility/test-api-manual/run-fake-test-framework.js:2]: Is this `require` necessary? You are doing a require with `--require` here: https://github.com/DataDog/dd-trace-js/pull/3202/files#diff-2d02c6c61a37789299433916e8654ada993c0f6b43f2a881c0ef8019f4227269R71

File: integration-tests/test-api-manual.spec.js [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+        cwd,
+        env: { ...getCiVisAgentlessConfig(receiver.port), DD_CIVISIBILITY_MANUAL_API_ENABLED: '1' },
+        stdio: 'pipe'
+      }
+    )
+    childProcess.on('exit', () => {
+      receiverPromise.then(() => done())
+    })
+  })
+  it('does not report test spans if DD_CIVISIBILITY_MANUAL_API_ENABLED is not set', (done) => {
```
Comment (uurien) [inline integration-tests/test-api-manual.spec.js:83]:
```suggestion

  it('does not report test spans if DD_CIVISIBILITY_MANUAL_API_ENABLED is not set', (done) => {
```

File: packages/dd-trace/src/config.js [hunk 11]
```diff
@@ -575,6 +580,7 @@ ken|consumer_?(?:id|key|secret)|sign(?:ed|ature)?|auth(?:entication|orization)?)
       (this.isIntelligentTestRunnerEnabled && !isFalse(DD_CIVISIBILITY_GIT_UPLOAD_ENABLED))
 
     this.gitMetadataEnabled = isTrue(DD_TRACE_GIT_METADATA_ENABLED)
+    this.isManualApiEnabled = this.isCiVisibility && isTrue(DD_CIVISIBILITY_MANUAL_API_ENABLED)
```
Comment (uurien) [inline packages/dd-trace/src/config.js:583]: I think that we should add the test in `config.spec.js`

File: integration-tests/ci-visibility/test-api-manual/run-fake-test-framework.js (hunk 8 above)
Comment (juan-fernandez) [reply integration-tests/ci-visibility/test-api-manual/run-fake-test-framework.js:2]: yeah that's a good point. Likely not

File: packages/dd-trace/src/config.js (hunk 11 above)
Comment (juan-fernandez) [reply packages/dd-trace/src/config.js:583]: indeed, good catch

Comment (juan-fernandez) [reply packages/dd-trace/src/config.js:583]: added a test in https://github.com/DataDog/dd-trace-js/pull/3202/commits/d7768fa1687dd12d86033ddc978e219c154fafca

File: integration-tests/ci-visibility/test-api-manual/run-fake-test-framework.js (hunk 8 above)
Comment (juan-fernandez) [reply integration-tests/ci-visibility/test-api-manual/run-fake-test-framework.js]: removed it in https://github.com/DataDog/dd-trace-js/pull/3202/commits/d7768fa1687dd12d86033ddc978e219c154fafca

Comment (juan-fernandez) [reply integration-tests/ci-visibility/test-api-manual/run-fake-test-framework.js]: done in https://github.com/DataDog/dd-trace-js/pull/3202/commits/d7768fa1687dd12d86033ddc978e219c154fafca

**Review by uurien — APPROVED** [review]

**Review by rochdev — APPROVED** [review]
